### PR TITLE
Add Acronym Support for Fuzzy Search

### PR DESCRIPTION
**Pull request by taooceros**
1. This pull request add acronym support for fuzzysearch. Link to #136 
2. This pr adds mapping from translated text to original text, allowing better highlighted text display for people using pinyin search (or potential future translation).

### CODE REVIEW

**Comment by taooceros** [comment]
I originally want to do it as the way I have done before, but then I change my mind and decide to write a straight-forward search for the acronyms and query, which will make retrieve metadata easier and support the case capital letter and space both exist.

Because it is a straightforward loop, the performance should be good.

**Comment by taooceros** [comment]
I think the test case need update

**Comment by taooceros** [comment]
If this pull request has been approved, #183 should remove the part that create acronym before the pinyin. However, in order to achieve a correct matchdata highlight, IAlphabet should return an extra bool value suggesting whether the string is translated, since one Chinese character will project to multiple letter in pinyin form.

**Comment by taooceros** [comment]
I am not sure how we should implement the data match for translated data, since the length change in translation...... should we simply don't highlight data when the string has been translated? Or return an extra string for matching data, which make it the same length as the original string.
Or create an extra class for mapping the index of translated string to original string, and create highlighted index based on that.

**Comment by taooceros** [comment]
@jjw24 please take a try on the new mapping class, and it allow us to even highlight text when using full pinyin. However, it will take another loop to do the mapping every time, not sure if it will be possible to optimize it more.

**Comment by jjw24** [comment]
just need some clarification:
- is this acronym support for cases where 'vsc' gets 'visual studio code'?
- essentially it is for identifying first letter of every word?

**Comment by taooceros** [comment]
> just need some clarification:
> - is this acronym support for cases where 'vsc' gets 'visual studio code'?
> - essentially it identifies first letter of every word?
> - should 'vs' for 'visual studio code' match?
> - currently this pr even match 'v' for 'visual studio code' which is a bit excessive

It matches the first letter of a word seperated from whitespace or upper letter. So vsc will match both "visual studio code" or "Visual Studio Code" or "visualStudioCode" or "VisualStudioCode".
Not sure whether we should make vs match for visual studio code. I think you are right that a small panalty should be added for unmatched after query finish, while a larger panalty has been added for unmatches before query end.

**Comment by jjw24** [comment]
When would there be a scenario where "visualStudioCode" or "VisualStudioCode" need to be acronym matched? program names are usually seperated by space, and if not the case it think should leave it to the fuzzy search to match up those kind of strings, that's what fuzzy matching is for.

I am concerned with performance when you have that loop at the beginning to do acronym search and you iterate through the whole 'stringtocompare' if that does not get a match, the rest of the code does another loop through the 'stringtocompare'.

i suggest we work it into the current loop. Take a look at the branch i pushed up:
https://github.com/Flow-Launcher/Flow.Launcher/compare/add_acronym_search?expand=1
https://github.com/Flow-Launcher/Flow.Launcher/tree/add_acronym_search

this has worked acronym search into the loop without another extra loop, it basically just projects forward to see the next substring if there is a match in case the loop is satisfied with the fuzzy search match and jumps to the next char.

I think it's a start, although it doesnt do the "visualStudioCode" or "VisualStudioCode" (which i think we should improve the actual fuzzy search code to do this) or where the the substring is numerical.

**Comment by taooceros** [comment]
> When would there be a scenario where "visualStudioCode" or "VisualStudioCode" need to be acronym matched? program names are usually seperated by space, and if not the case it think should leave it to the fuzzy search to match up those kind of strings, that's what fuzzy matching is for.
> 
> I am concerned with performance when you have that loop at the beginning to do acronym search and you iterate through the whole 'stringtocompare' if that does not get a match, the rest of the code does another loop through the 'stringtocompare'.
> 
> i suggest we work it into the current loop. Take a look at the branch i pushed up:
> https://github.com/Flow-Launcher/Flow.Launcher/compare/add_acronym_search?expand=1
> (https://github.com/Flow-Launcher/Flow.Launcher/tree/add_acronym_search)
> 
> this has worked acronym search into the loop without another extra loop, it basically just projects forward to see if there is a match in the next substring in case the loop is satisfied with the fuzzy search match and jumps to the next char.
> 
> I think it's a start, although it doesnt do the "visualStudioCode" or "VisualStudioCode" (which i think we should improve the actual fuzzy search code to do this) or where the the substring is numerical.
> 
> Let me know what you think of this change

I added that because in my computer sometimes programs split there name like CamelCase or Pascal Case, for example "PostmanCanary" or "JetBrains". So I added that feature.

**Comment by taooceros** [comment]
I will take a look on the branch, soon. However, I have encountered an issue that after merging dev branch, the setting windows is unable to get the dynamic resource, which most of the text in textblock disappears. I am not sure what cause that.....

**Comment by jjw24** [comment]
> I will take a look on the branch, soon. However, I have encountered an issue that after merging dev branch, the setting windows is unable to get the dynamic resource, which most of the text in textblock disappears. I am not sure what cause that.....

try delete output folder and rebuild

>I added that because in my computer sometimes programs split there name like CamelCase or Pascal Case, for example "PostmanCanary" or "JetBrains". So I added that feature.

hmm fair point, let me have a think how we can build it in.

**Comment by taooceros** [comment]
> > I will take a look on the branch, soon. However, I have encountered an issue that after merging dev branch, the setting windows is unable to get the dynamic resource, which most of the text in textblock disappears. I am not sure what cause that.....
> 
> try delete output folder and rebuild

Let me take a try.

> 
> > I added that because in my computer sometimes programs split there name like CamelCase or Pascal Case, for example "PostmanCanary" or "JetBrains". So I added that feature.
> 
> hmm fair point, let me have a think how we can build it in.

I have checked the code for split right now, and it require an extra for loop to do that. So it will be the same for us to use our for loop to do the same jobs. However, I do think we can merge fuzzy search and acronym search since fuzzy search do perform a for loop. I don't do that because I haven't fully understand the full fuzzy search, so it make me a few scare to touch some important things. Let me take a try on that.

**Comment by taooceros** [comment]
By the way, Marry Christmas! ❤

**Comment by taooceros** [comment]
> try delete output folder and rebuild

That seems fix it. Thank you

**Comment by jjw24** [comment]
> By the way, Marry Christmas! ❤

Haha thank you. Merry Christmas to you too :)

**Comment by jjw24** [comment]
> I don't do that because I haven't fully understand the full fuzzy search, so it make me a few scare to touch some important things. Let me take a try on that.

No stress, we have a lot of tests around it.

>I have checked the code for split right now, and it require an extra for loop to do that.

Yeah I wonder if there is a way to do it without the split

**Comment by jjw24** [comment]
please resolve conflict @taooceros

**Comment by taooceros** [comment]
Sorry I merge the wrong one, I will reset it soon.

**Comment by taooceros** [comment]
ok, this should be great now...😅 @jjw24

**Comment by jjw24** [comment]
No problem will take it for a spin

File: Flow.Launcher.Infrastructure/StringMatcher.cs
Comment (jjw24) [inline Flow.Launcher.Infrastructure/StringMatcher.cs:189]: this change is correct?

File: Flow.Launcher.Infrastructure/StringMatcher.cs
Comment (jjw24) [inline Flow.Launcher.Infrastructure/StringMatcher.cs:43]: please update the FuzzyMatch summary to include the Acronym search steps

File: Flow.Launcher.Infrastructure/StringMatcher.cs
Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:189]: Em yes, that is used to before early return of full match to check more acronym.

File: Flow.Launcher.Infrastructure/StringMatcher.cs
Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: 👌, and I have a question, should we add an option to open acronym search?

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: sorry what do you mean open acronym search?

Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: I mean we can have a parameters that indicates whether the search should include acronym search (maybe default is true).

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: any reason why someone wouldnt use it or have the search method included?

Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: Emmm I am not quite sure. For example, maybe we don't want acronym search used to matching for explorer plugin because the index search doesn't include acronym search.
Maybe for some description matching (generally in subtitle matching), we don't want acronym search.

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: Good point, though Explorer's search can't utilize fuzzy search at all due to usage of DirectiryInfoSearch and Index, so no need to worry here. 

If we don't want description search might as well not use fuzzy on it either.

Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: Well, I mean the Explorer utilize fuzzy search to create highlight data for results in result manager, so that may be a place that shouldn't use acronym search. ```

```

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: but you can use acronym search with Explorer, so you cant type 'gk' and hit GitKraken, you have to type 'git' so highlight data would still be correct right?

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [inline Flow.Launcher.Test/FuzzyMatcherTest.cs:152]: why is this test case removed?

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [inline Flow.Launcher.Test/FuzzyMatcherTest.cs:134]: this shouldnt have such a high score that it shows, acronym search matches it despite 'Help' at the start?

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [inline Flow.Launcher.Test/FuzzyMatcherTest.cs:137]: this score has dipped due to acronym penalty?

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [inline Flow.Launcher.Test/FuzzyMatcherTest.cs:157]: why has this precision changed to regular?

File: Flow.Launcher.Infrastructure/StringMatcher.cs
Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:43]: Yeah I agree. I think there is some conner case but it is not a lot.

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [inline Flow.Launcher.Test/FuzzyMatcherTest.cs:205]: hmm ok, so acronym match doesnt have to match entire string

File: Flow.Launcher/ViewModel/MainViewModel.cs
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:139]: what's this change?

**Review by jjw24 — CHANGES_REQUESTED** [review]
please update summary

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (taooceros) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:205]: Well, yes, for now. Because we include description in searching for now, which means that there will be quite a lot of unmatched acronym if we match the entire string.

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:134]: hmm despite the penalty, i wonder if this is will bring up irrelevant results

File: Flow.Launcher/ViewModel/MainViewModel.cs
Comment (taooceros) [reply Flow.Launcher/ViewModel/MainViewModel.cs:139]: That's due to unintended merge. It seems just swap the posible of the two registration.

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (taooceros) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:152]: Because the acronym search will make this case matched, through `cure hope raise`. I think it is not the original idea for this test?

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (taooceros) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:157]: Because the acronym search will allow that to be matched with SearchPrecisionScore.Regular

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (jjw24) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:152]: This was to prevent fuzzy match bring up this kind of unrelated result. Maybe scoring needs adjusting so three acronyms shouldn't being up this long string?

File: Flow.Launcher/ViewModel/MainViewModel.cs
Comment (jjw24) [reply Flow.Launcher/ViewModel/MainViewModel.cs:139]: There is a bit of difference, can you please double check including it is ok

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (taooceros) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:137]: yes, it match the middle "SQL" but without the M

**Comment by jjw24** [comment]
Feels like we should adjust the scoring. With the current code, is it possible to score based on percentage of acronyms matched? So three characters is not going to bring up an entire long phrase

**Comment by taooceros** [comment]
> Feels like we should adjust the scoring. With the current code, is it possible to score based on percentage of acronyms matched? So three characters is not going to bring up an entire long phrase

Yes, I agree with you. However, I think we still match the description in Program plugin, which means some description is long? Maybe we shall seperate them, with matching title with acronym and description without acronym, and add panelty for mismatch after the query is finished for acronym.

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (taooceros) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:134]: I am not quite sure, but there shouldn't be few strings that will act like that. Maybe we can release a pre-release and check how people respond to it.

File: Flow.Launcher.Test/FuzzyMatcherTest.cs
Comment (taooceros) [reply Flow.Launcher.Test/FuzzyMatcherTest.cs:152]: emmm agree with you.

**Comment by taooceros** [comment]
> please update summary

Will do tmr, or on the weekend.

**Comment by jjw24** [comment]
> > Feels like we should adjust the scoring. With the current code, is it possible to score based on percentage of acronyms matched? So three characters is not going to bring up an entire long phrase
> 
> Yes, I agree with you. However, I think we still match the description in Program plugin, which means some description is long? Maybe we shall seperate them, with matching title with acronym and description without acronym, and add panelty for mismatch after the query is finished for acronym.
> 
> # e.g.
> image

i dont think the problem is whether we match description or not, in general acronym shouldnt bring up irrelevant long strings. That's why i was thinking maybe scoring needs to be percentage based. I dont think we should specifically exclude acronym from description because this will start to get confusing, not to mention the string matcher class is getting dangerously unmaintainable. 

What you have with penalty may work as well, and for long strings penalty could be so large that the acronym score will go into negative, which is fine as users can still use rest of fuzzy search method to get the results if they were to look for a long string.

**Comment by jjw24** [comment]
Just reworking scoring to be based on percentage not total acronyms so code doesn't bring up long strings. Will push up the branch for you to take a look

**Comment by taooceros** [comment]
Thank you!

File: Flow.Launcher.Infrastructure/StringMatcher.cs
Comment (jjw24) [inline Flow.Launcher.Infrastructure/StringMatcher.cs:107]: @taooceros what's this 'spaceMet' variable trying to solve and is this needed for translation?

Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:107]: I think it is used to mark character after space to be acronym even when it is lower case.

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:107]: Ok cool, I thought so as well. I am thinking of using substring index = 0 to replace this, since the start of substring char is always acronym

Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:107]: Have we split the string to compare? I think we only split the query string.

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:107]: Oh we didn't, lol ok nvm then

Comment (jjw24) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:107]: Alternatively then, we can go back an index to check if it is space

Comment (taooceros) [reply Flow.Launcher.Infrastructure/StringMatcher.cs:107]: Yeah, that's feasible.

**Comment by jjw24** [comment]
please look at the updates and let me know if good to merge in here:
https://github.com/taooceros/Flow.Launcher/pull/4/files

**Comment by taooceros** [comment]
> please look at the updates and let me know if good to merge in here:
> https://github.com/taooceros/Flow.Launcher/pull/4/files

Great! I will take a look soon.

**Comment by taooceros** [comment]
@jjw24 any further change we need to make for this PR?

**Comment by jjw24** [comment]
Let me take a final look in 2hrs

**Review by jjw24 — APPROVED** [review]
Great work @taooceros 👍

**Pull request merged**
